### PR TITLE
Bad word filter Feature (Ready to Use) :D

### PR DESCRIPTION
**Pull request by capensisma**
# Bad Words Filter (Feature)
Here is the bad words filter feature where you can filter all the bad words : 

- No need to add them just Enable/Disable in Settings-Message (default = enabled)
- You can any words to filter and as much as you want ;)
- It's Awesome you can try with some phrases like `Asshole motherfucker with a big butt and nice pussy and I hate your fucking shit` Only for test purposes :D

Try and enjoy :)

### CODE REVIEW

**Comment by capensisma** [comment]
result would be `******* ************ with a big **** and nice ***** and I hate your ******* ****`

File: packages/rocketchat-lib/server/methods/filterBadWords.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+                var badWordsList = RocketChat.settings.get('Message_BadWordsFilterList');
+                    // console.log(badWordsList)
+                    //Add words to the blacklist
+                    if(!!badWordsList && badWordsList.length)
+                        {   
+                            // console.log(badWordsList)
+                            var filter = new Filter({ list: badWordsList.split(',') });
+                        }
+                    else
+                        var filter = new Filter();
```
Comment (rodrigok) [inline packages/rocketchat-lib/server/methods/filterBadWords.js]: Codacy Issue found: [Expected '{' and instead saw 'var'.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2426118531/issues/source?bid=3199634&fileBranchId=3240346#l19)

Comment (rodrigok) [inline packages/rocketchat-lib/server/methods/filterBadWords.js]: Codacy Issue found: [Variable 'filter' is already declared in line 16.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2426118531/issues/source?bid=3199634&fileBranchId=3240346#l19)

File: packages/rocketchat-lib/server/methods/filterBadWords.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+                
+                var badWordsList = RocketChat.settings.get('Message_BadWordsFilterList');
+                    // console.log(badWordsList)
+                    //Add words to the blacklist
+                    if(!!badWordsList && badWordsList.length)
+                        {   
+                            // console.log(badWordsList)
+                            var filter = new Filter({ list: badWordsList.split(',') });
+                        }
+                    else
```
Comment (rodrigok) [inline packages/rocketchat-lib/server/methods/filterBadWords.js]: Codacy Issue found: [Expected { after 'else'.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2426118531/issues/source?bid=3199634&fileBranchId=3240346#l18)

File: packages/rocketchat-lib/server/methods/filterBadWords.js (hunk 1 above)
Comment (rodrigok) [inline packages/rocketchat-lib/server/methods/filterBadWords.js]: Codacy Issue found: ["filter" is already defined](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2426118531/issues/source?bid=3199634&fileBranchId=3240346#l19)

File: packages/rocketchat-lib/server/methods/filterBadWords.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+                        var filter = new Filter();
+                message.msg = filter.clean(message.msg);
```
Comment (rodrigok) [inline packages/rocketchat-lib/server/methods/filterBadWords.js]: Codacy Issue found: ["filter" used outside of binding context.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2426118531/issues/source?bid=3199634&fileBranchId=3240346#l20)

File: packages/rocketchat-lib/server/methods/filterBadWords.js (hunk 1 above)
Comment (rodrigok) [inline packages/rocketchat-lib/server/methods/filterBadWords.js]: Codacy Issue found: ["filter" used outside of binding context.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2426118531/issues/source?bid=3199634&fileBranchId=3240346#l19)

**Comment by capensisma** [comment]
@rodrigok it works and I tested it many times bro ;)

**Comment by engelgabriel** [comment]
@capensisma this is an automated test o code quality, from https://www.codacy.com/app/RocketChat/Rocket-Chat/pullRequest?prid=220222

**Comment by capensisma** [comment]
@engelgabriel there is nothing to be improved there, so ... and it works well and as intended

**Comment by capensisma** [comment]
CI is good though https://travis-ci.org/RocketChat/Rocket.Chat/builds/124090832

**Comment by engelgabriel** [comment]
It may work well.. this is not the point. 

The code QUALITY is bad because it doens't follow the standards.

**Comment by capensisma** [comment]
code QUALITY is bad? okay show me better then!

**Comment by engelgabriel** [comment]
What editor do you use? Can you install ESlint?

**Comment by capensisma** [comment]
I can in the cloud c9.io that's all why is the editor related to this?

**Comment by rodrigok** [comment]
@capensisma a better code syntax looks like:

```javascript
const Filter = Npm.require('bad-words');

RocketChat.callbacks.add('beforeSaveMessage', function(message) {

	if (RocketChat.settings.get('Message_AllowBadWordsFilter')) {
		const badWordsList = RocketChat.settings.get('Message_BadWordsFilterList');

		let filter;

		//Add words to the blacklist
		if (!!badWordsList && badWordsList.length) {
			filter = new Filter({ list: badWordsList.split(',') });
		} else {
			filter = new Filter();
		}

		message.msg = filter.clean(message.msg);
	}

	return message;
}, 1);
```

File: packages/rocketchat-lib/server/methods/filterBadWords.js (hunk 1 above)
Comment (rodrigok) [inline packages/rocketchat-lib/server/methods/filterBadWords.js]: Codacy Issue found: [Variable 'filter' is already declared in line 16.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2432869137/issues/source?bid=3199634&fileBranchId=3240346#l19)

Comment (rodrigok) [inline packages/rocketchat-lib/server/methods/filterBadWords.js]: Codacy Issue found: [Expected '{' and instead saw 'var'.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2432869137/issues/source?bid=3199634&fileBranchId=3240346#l19)

File: packages/rocketchat-lib/server/methods/filterBadWords.js [hunk 4]
```diff
@@ -0,0 +1,25 @@
+
+var Filter = Npm.require('bad-words');
+    
+    RocketChat.callbacks.add( 'beforeSaveMessage' , function(message){
+        // console.log(message)
+        
+        if( RocketChat.settings.get('Message_AllowBadWordsFilter') )
```
Comment (rodrigok) [inline packages/rocketchat-lib/server/methods/filterBadWords.js]: Codacy Issue found: [Keyword "if" must be followed by whitespace.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2432869137/issues/source?bid=3199634&fileBranchId=3240346#l7)

File: packages/rocketchat-lib/server/methods/filterBadWords.js [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+                            var filter = new Filter({ list: badWordsList.split(',') });
+                        }
+                    else
+                        var filter = new Filter();
+                message.msg = filter.clean(message.msg);
+                
+            }
+    return message;
+    
+    },1);
```
Comment (rodrigok) [inline packages/rocketchat-lib/server/methods/filterBadWords.js]: Codacy Issue found: [A space is required after ','.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2432869137/issues/source?bid=3199634&fileBranchId=3240346#l25)

File: packages/rocketchat-lib/server/methods/filterBadWords.js (hunk 2 above)
Comment (rodrigok) [inline packages/rocketchat-lib/server/methods/filterBadWords.js]: Codacy Issue found: [Expected { after 'else'.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2432869137/issues/source?bid=3199634&fileBranchId=3240346#l18)

File: packages/rocketchat-lib/server/methods/filterBadWords.js [hunk 6]
```diff
@@ -0,0 +1,25 @@
+
+var Filter = Npm.require('bad-words');
+
```
Comment (rodrigok) [inline packages/rocketchat-lib/server/methods/filterBadWords.js]: Codacy Issue found: [Trailing spaces not allowed.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2432869137/issues/source?bid=3199634&fileBranchId=3240346#l3)

File: packages/rocketchat-lib/server/methods/filterBadWords.js [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+            {
+                
+                var badWordsList = RocketChat.settings.get('Message_BadWordsFilterList');
+                    // console.log(badWordsList)
+                    //Add words to the blacklist
+                    if(!!badWordsList && badWordsList.length)
+                        {   
+                            // console.log(badWordsList)
+                            var filter = new Filter({ list: badWordsList.split(',') });
+                        }
```
Comment (rodrigok) [inline packages/rocketchat-lib/server/methods/filterBadWords.js]: Codacy Issue found: [Expected indentation of 20 space characters but found 24.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2432869137/issues/source?bid=3199634&fileBranchId=3240346#l17)

File: packages/rocketchat-lib/server/methods/filterBadWords.js (hunk 4 above)
Comment (rodrigok) [inline packages/rocketchat-lib/server/methods/filterBadWords.js]: Codacy Issue found: [Expected indentation of 6 space characters but found 8.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2432869137/issues/source?bid=3199634&fileBranchId=3240346#l7)

File: packages/rocketchat-lib/server/methods/filterBadWords.js [hunk 8]
```diff
@@ -0,0 +1,25 @@
+
+var Filter = Npm.require('bad-words');
+    
+    RocketChat.callbacks.add( 'beforeSaveMessage' , function(message){
```
Comment (rodrigok) [inline packages/rocketchat-lib/server/methods/filterBadWords.js]: Codacy Issue found: [There should be no space before ','.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2432869137/issues/source?bid=3199634&fileBranchId=3240346#l4)

File: packages/rocketchat-lib/server/methods/filterBadWords.js [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+        // console.log(message)
+        
+        if( RocketChat.settings.get('Message_AllowBadWordsFilter') )
+            {
+                
+                var badWordsList = RocketChat.settings.get('Message_BadWordsFilterList');
+                    // console.log(badWordsList)
+                    //Add words to the blacklist
+                    if(!!badWordsList && badWordsList.length)
```
Comment (rodrigok) [inline packages/rocketchat-lib/server/methods/filterBadWords.js]: Codacy Issue found: [Expected indentation of 10 space characters but found 20.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2432869137/issues/source?bid=3199634&fileBranchId=3240346#l13)

File: packages/rocketchat-lib/server/methods/filterBadWords.js (hunk 1 above)
Comment (rodrigok) [inline packages/rocketchat-lib/server/methods/filterBadWords.js]: Codacy Issue found: [Expected '{' and instead saw 'var'.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2433000388/issues/source?bid=3199634&fileBranchId=3240346#l19)

Comment (rodrigok) [inline packages/rocketchat-lib/server/methods/filterBadWords.js]: Codacy Issue found: [Variable 'filter' is already declared in line 16.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2433000388/issues/source?bid=3199634&fileBranchId=3240346#l19)

File: packages/rocketchat-lib/server/methods/filterBadWords.js (hunk 4 above)
Comment (rodrigok) [inline packages/rocketchat-lib/server/methods/filterBadWords.js]: Codacy Issue found: [Keyword "if" must be followed by whitespace.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2433000388/issues/source?bid=3199634&fileBranchId=3240346#l7)

File: packages/rocketchat-lib/server/methods/filterBadWords.js (hunk 5 above)
Comment (rodrigok) [inline packages/rocketchat-lib/server/methods/filterBadWords.js]: Codacy Issue found: [A space is required after ','.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2433000388/issues/source?bid=3199634&fileBranchId=3240346#l25)

File: packages/rocketchat-lib/server/methods/filterBadWords.js (hunk 2 above)
Comment (rodrigok) [inline packages/rocketchat-lib/server/methods/filterBadWords.js]: Codacy Issue found: [Expected { after 'else'.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2433000388/issues/source?bid=3199634&fileBranchId=3240346#l18)

File: packages/rocketchat-lib/server/methods/filterBadWords.js (hunk 6 above)
Comment (rodrigok) [inline packages/rocketchat-lib/server/methods/filterBadWords.js]: Codacy Issue found: [Trailing spaces not allowed.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2433000388/issues/source?bid=3199634&fileBranchId=3240346#l3)

File: packages/rocketchat-lib/server/methods/filterBadWords.js (hunk 8 above)
Comment (rodrigok) [inline packages/rocketchat-lib/server/methods/filterBadWords.js]: Codacy Issue found: [There should be no space before ','.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2433000388/issues/source?bid=3199634&fileBranchId=3240346#l4)

File: packages/rocketchat-lib/server/methods/filterBadWords.js (hunk 1 above)
Comment (rodrigok) [inline packages/rocketchat-lib/server/methods/filterBadWords.js]: Codacy Issue found: ["filter" is already defined](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2433000388/issues/source?bid=3199634&fileBranchId=3240346#l19)

File: packages/rocketchat-lib/server/methods/filterBadWords.js [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+                        {
```
Comment (rodrigok) [inline packages/rocketchat-lib/server/methods/filterBadWords.js]: Codacy Issue found: [Trailing spaces not allowed.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2433000388/issues/source?bid=3199634&fileBranchId=3240346#l14)

File: packages/rocketchat-lib/server/methods/filterBadWords.js (hunk 8 above)
Comment (rodrigok) [inline packages/rocketchat-lib/server/methods/filterBadWords.js]: Codacy Issue found: [Missing space before opening brace.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2433000388/issues/source?bid=3199634&fileBranchId=3240346#l4)

File: packages/rocketchat-lib/server/methods/filterBadWords.js (hunk 1 above)
Comment (rodrigok) [inline packages/rocketchat-lib/server/methods/filterBadWords.js]: Codacy Issue found: [Expected '{' and instead saw 'var'.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2433000388/issues/source?bid=3199634&fileBranchId=3240346#l19)

Comment (rodrigok) [inline packages/rocketchat-lib/server/methods/filterBadWords.js]: Codacy Issue found: [Variable 'filter' is already declared in line 16.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2433000388/issues/source?bid=3199634&fileBranchId=3240346#l19)

File: packages/rocketchat-lib/server/methods/filterBadWords.js (hunk 4 above)
Comment (rodrigok) [inline packages/rocketchat-lib/server/methods/filterBadWords.js]: Codacy Issue found: [Keyword "if" must be followed by whitespace.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2433000388/issues/source?bid=3199634&fileBranchId=3240346#l7)

File: packages/rocketchat-lib/server/methods/filterBadWords.js (hunk 5 above)
Comment (rodrigok) [inline packages/rocketchat-lib/server/methods/filterBadWords.js]: Codacy Issue found: [A space is required after ','.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2433000388/issues/source?bid=3199634&fileBranchId=3240346#l25)

File: packages/rocketchat-lib/server/methods/filterBadWords.js (hunk 2 above)
Comment (rodrigok) [inline packages/rocketchat-lib/server/methods/filterBadWords.js]: Codacy Issue found: [Expected { after 'else'.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2433000388/issues/source?bid=3199634&fileBranchId=3240346#l18)

File: packages/rocketchat-lib/server/methods/filterBadWords.js (hunk 6 above)
Comment (rodrigok) [inline packages/rocketchat-lib/server/methods/filterBadWords.js]: Codacy Issue found: [Trailing spaces not allowed.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2433000388/issues/source?bid=3199634&fileBranchId=3240346#l3)

File: packages/rocketchat-lib/server/methods/filterBadWords.js (hunk 8 above)
Comment (rodrigok) [inline packages/rocketchat-lib/server/methods/filterBadWords.js]: Codacy Issue found: [There should be no space before ','.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2433000388/issues/source?bid=3199634&fileBranchId=3240346#l4)

File: packages/rocketchat-lib/server/methods/filterBadWords.js (hunk 1 above)
Comment (rodrigok) [inline packages/rocketchat-lib/server/methods/filterBadWords.js]: Codacy Issue found: ["filter" is already defined](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2433000388/issues/source?bid=3199634&fileBranchId=3240346#l19)

File: packages/rocketchat-lib/server/methods/filterBadWords.js (hunk 10 above)
Comment (rodrigok) [inline packages/rocketchat-lib/server/methods/filterBadWords.js]: Codacy Issue found: [Trailing spaces not allowed.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2433000388/issues/source?bid=3199634&fileBranchId=3240346#l14)

File: packages/rocketchat-lib/server/methods/filterBadWords.js (hunk 8 above)
Comment (rodrigok) [inline packages/rocketchat-lib/server/methods/filterBadWords.js]: Codacy Issue found: [Missing space before opening brace.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2433000388/issues/source?bid=3199634&fileBranchId=3240346#l4)

**Comment by capensisma** [comment]
What's the ETA on this? will never be merged or what? bros?

**Comment by geekgonecrazy** [comment]
@capensisma can you take a look at the few code tweaks as included in the previous comment, and the comments on the commit?

All are listed here above.  We review carefully anything that will directly interact with the messages.  We want to make sure to prevent any issues from coming up.  Fixing the above mentioned would give us more comfort :)

Appreciate the understanding!

**Comment by capensisma** [comment]
I swear to God there is nothing to change or fix here, its just the codacy doesn't like the indentations or something and it's says `var ` or like ` Keyword "if" must be followed by whitespace.` this is code quality for CoffeeScript not Pure Javascript bros C'mon!

**Comment by geekgonecrazy** [comment]
@capensisma The thing you have to remember about code you contribute.  Once you check it in to the project and we accept it, you no longer have to maintain it.  We do.  Its because of this we are trying to improve code quality.  

But no worries.. I'll take care of this.  Thanks for contributing.

**Comment by geekgonecrazy** [comment]
We'll review and merge via #2991

**Pull request closed without merging**